daemons/dss-sync/app.py

```diff
@@ -24,7 +33,97 @@ def process_new_syncable_object(event, context):
     if event.get("Event") == "s3:TestEvent":
         app.log.info("DSS sync daemon received S3 test event")
     else:
-        s3 = boto3.resource("s3")
-        bucket = s3.Bucket(event['Records'][0]["s3"]["bucket"]["name"])
+        bucket = resources.s3.Bucket(event['Records'][0]["s3"]["bucket"]["name"])
         obj = bucket.Object(unquote(event['Records'][0]["s3"]["object"]["key"]))
-        sync_blob(source_platform="s3", source_key=obj.key, dest_platform="gs", logger=app.log)
+        sync_blob(source_platform="s3", source_key=obj.key, dest_platform="gs", logger=app.log, context=context)
+
+@app.sns_topic_subscriber("dss-gs-composite-upload-ready")
+def compose_upload(event, context):
+    """
+    See https://cloud.google.com/storage/docs/composite-objects for details of the Google Storage API used here.
+    """
+    gs_max_compose_parts = 32
+    msg = json.loads(event["Records"][0]["Sns"]["Message"])
+    gs = dss.Config.get_cloud_specific_handles("gcp")[0].gcp_client
+    gs_bucket = gs.get_bucket(msg["dest_bucket"])
+    while True:
+        try:
+            context.log("Composing, stage 1")
+            compose_stage2_blob_names = []
+            if msg["total_parts"] > gs_max_compose_parts:
+                for part_id in range(1, msg["total_parts"] + 1, gs_max_compose_parts):
```

Comment:
S3 multipart upload part IDs are integers between [1, 10000]. Although GS part IDs are freeform string keys, the enumeration here is a byproduct of the fact that part IDs are generated in `dispatch_multipart_sync()` without regard to which platform is the destination.